network-area-diagram-viewer/src/main/resources/svg.js

```diff
@@ -0,0 +1,463 @@
+            transform.setTranslate(translation.x, translation.y);
+            if (CONTINUOUS_UPDATE) {
+                diagram.update(selectedElement, position, translation);
+            }
+        }
+    }
+
+    function endDrag(event) {
+        if (selectedElement) {
+            console.log("endDrag for element " + selectedElement.id);
+            selectedElement.style.cursor = 'grab';
+            // Desired position for the selected element is current mouse position in SVG coordinate system
+            var position = getMousePosition(event);
+            var translation = {x: position.x - offset.x, y: position.y - offset.y};
+            diagram.update(selectedElement, position, translation);
+        }
+        selectedElement = null;
+    }
+}
+
+// PowSyBl Diagram
+
+function Diagram(svg) {
+    this.getDraggableFrom = getDraggableFrom;
+    this.update = update;
+
+    function getDraggableFrom(event) {
+        if (isDraggable(event.target)) {
+            return event.target;
+        } else if (isDraggable(event.target.parentElement)) {
+            return event.target.parentElement;
+        }
+    }
+
+    function isDraggable(element) {
+        return hasId(element) && classIsDraggable(element);
+    }
+
+    function classIsDraggable(element) {
+        // For svg objects it is better to get class names through attribute, to avoid potential SVGAnimatedString values
+        // https://developer.mozilla.org/en-US/docs/Web/API/Element/className
+        var className = element.getAttribute("class");
+        if (!className) {
+            var parent = element.parentElement;
+            if (parent) {
+                className = parent.getAttribute("class");
+            }
+        }
+        return className && (className.includes("vl-node") || className.includes("boundary-node") || className.includes("busnode"));
+    }
+
+    function hasId(element){
+        return typeof element.id != 'undefined' && element.id != '';
+    }
+
+    function update(svgElem, position, translation) {
+        var updateOnServer = false;
+        if (updateOnServer) {
+            updateOnServer();
+        } else {
+            // Adjust the id of the element that has been moved
+            // It we were moving a bus node (its circle),
+            // find the id of the corresponding voltage level
+            var id = svgElem.getAttribute("id");
+            var parent = svgElem.parentElement;
+            var idParent = parent.getAttribute("id");
+            if (idParent) {
+                id = idParent;
+            }
+            updateOnClient(id, position, translation);
+        }
+    }
+
+    function updateOnClient(id, position, translation) {
+        console.log("updateOnClient, moved node id = " + id);
+        translateText(id, translation);
+        // TODO(Luma) Update here the position of all inner busNodes
+        for (var edge of svg.getElementsByTagName("nad:edge")) {
```

Comment:
you don't need to write this loop, you should be able to use something like 
```document.querySelector("metadata nad:edges nad:edge[node1=xxx][node2=yyy]")```

Besides you're doing that on each `mouseMove`, maybe the interaction can be smoother by adding some "metadata" about the current selected element to avoid doing that all the time.

Reply:
Sure, it can be optimized. I will follow your suggestions

Reply:
done the first part: use query selector instead of looping through all the edges

Reply:
and also cached the results of the query so we do not perform it in every mouse move, only when the moved node changes